src/typeChecker/__tests__/typeChecker.test.ts

```diff
@@ -579,10 +600,485 @@ describe('type checking functions used in polymorphic fashion', () => {
+
+describe('type checking for loops', () => {
+  it('allows for correctly formed for loops with let outside and inside', () => {
+    const code = `
+      let a = false;
+      for (let a = 3; a < 5; a = a + 1) {
+        a + 20;
+      }
+    `
+    // console.log(parse(code, 3).body[1])
+    const [program, errors] = parseAndTypeCheck(code, 3)
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`"a: boolean"`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`""`)
+  })
+
+  it('allows for correctly formed for loops with let outside only', () => {
+    const code = `
+      let a = 200;
+      for (a = 3; a < 5; a = a + 1) {
+        a + 20;
+      }
+    `
+    // console.log(parse(code, 3).body[1])
+    const [program, errors] = parseAndTypeCheck(code, 3)
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`"a: number"`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`""`)
+  })
+
+  it('fails when test is not a boolean', () => {
+    const code = `
+      let a = 200;
+      for (a = 3; a + 5; a = a + 1) {
+        a + 20;
+      }
+    `
+    const [program, errors] = parseAndTypeCheck(code, 3)
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`"a: number"`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`
+      "Line 3: Expected the test part of the for statement:
+        for (...; a + 5; ...) { ... }
+      to have type boolean, but instead it is type:
+        number"
+    `)
+  })
+
+  it('fails when initialized variable not used correctly', () => {
+    const code = `
+      let a = false;
+      for (let a = 3; a < 5; a = a || false) {
+        a && a;
+      }
+    `
+    const [program, errors] = parseAndTypeCheck(code, 3)
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`"a: boolean"`)
+    expect(parseError(errors)).toMatchInlineSnapshot(`
+      "Line 3: A type mismatch was detected in the binary expression:
+        a || false
+      The binary operator (||) expected two operands with types:
+        boolean || T0
+      but instead it received two operands of types:
+        number || boolean
+      Line 4: A type mismatch was detected in the binary expression:
+        a && a
+      The binary operator (&&) expected two operands with types:
+        boolean && T0
+      but instead it received two operands of types:
+        number && number"
+    `)
+  })
+})
+
+describe('type checking arrays', () => {
+  it('handles empty arrays', () => {
+    const code = `
+      const arr = [];
+      const y = 1 + arr[1]; // runtime error but no type error
+      arr[0] = false;
+      const x = 1 + arr[1]; // type error
+    `
+    const [program, errors] = parseAndTypeCheck(code, 3)
+    expect(topLevelTypesToString(program)).toMatchInlineSnapshot(`
+      "arr: Array<number>
+      y: number
+      x: number"
+    `)
+    expect(parseError(errors)).toMatchInlineSnapshot(`
+      "Line 4: Array expected type: Array<number>
+          but got: boolean"
```

Comment:
This error seems rather confusing.  Array<number> is the type of the expected array but boolean is the type of the expected array _element_ 

Ideally we should be consistent and display either the array type of both or the element type of both